### ISSUE TITLE
Bad performance on my custom dataset.

### ISSUE THREAD

**Issue by zjwzcx** [CoinCheung/BiSeNet#230]
Hi!
I tried to implement BiseNetv2 on my custom dataset for distinguishing paths and background (2 labels). I modified the model setting mainly referring to [this blog](https://blog.csdn.net/CZW199418/article/details/110500273). I guess that I've done all the neccessary modifications to fit my dataset.

To make sure BiSeNetv2 can be trained to converge, I trained it on a small dataset with only one image in my training set. I kept the maximum number of iterations at 150000, and trained at a learning rate of [5e-3, 5e-4, 5e-5] respectively long enough (I stopped training when I was sure that the training loss was no longer falling) and considered that over-fitting has occurred)

However, I found that the visualization result of overfitting_model.pth on the only one image in my training set is really bad. The label and result is as shown below:
![left000001_gtFine_labelIds](https://user-images.githubusercontent.com/43678835/163721419-7fc9a1a9-a4cc-48e3-9e26-2389bc3dc9e1.png)
![image_04_17_21_19](https://user-images.githubusercontent.com/43678835/163721423-b3a2ca18-7fa9-4655-acbe-53292bba0d47.png)

**Comment by zjwzcx** [comment]
The last few lines in .log file is as follows:
**
INFO logger.py(52): iter: 44650/150000, lr: 0.002013, eta: 2 days, 0:17:47, time: 83.23, loss: 0.00030681, loss_pre: 0.00000183, loss_aux0: 0.00017385, loss_aux1: 0.00010247, loss_aux2: 0.00002842, loss_aux3: 0.00000023
INFO logger.py(52): iter: 44700/150000, lr: 0.002012, eta: 2 days, 0:16:27, time: 83.40, loss: 0.00030099, loss_pre: 0.00000179, loss_aux0: 0.00017006, loss_aux1: 0.00009998, loss_aux2: 0.00002877, loss_aux3: 0.00000038
INFO logger.py(52): iter: 44750/150000, lr: 0.002011, eta: 2 days, 0:15:06, time: 83.31, loss: 0.00030304, loss_pre: 0.00000094, loss_aux0: 0.00017255, loss_aux1: 0.00010013, loss_aux2: 0.00002915, loss_aux3: 0.00000027
INFO logger.py(52): iter: 44800/150000, lr: 0.002010, eta: 2 days, 0:13:46, time: 83.51, loss: 0.00030106, loss_pre: 0.00000083, loss_aux0: 0.00017110, loss_aux1: 0.00009896, loss_aux2: 0.00002945, loss_aux3: 0.00000072
INFO logger.py(52): iter: 44850/150000, lr: 0.002010, eta: 2 days, 0:12:25, time: 83.14, loss: 0.00030173, loss_pre: 0.00000095, loss_aux0: 0.00017161, loss_aux1: 0.00010035, loss_aux2: 0.00002839, loss_aux3: 0.00000043
INFO logger.py(52): iter: 44900/150000, lr: 0.002009, eta: 2 days, 0:11:04, time: 83.54, loss: 0.00029890, loss_pre: 0.00000132, loss_aux0: 0.00017022, loss_aux1: 0.00009874, loss_aux2: 0.00002826, loss_aux3: 0.00000037
INFO logger.py(52): iter: 44950/150000, lr: 0.002008, eta: 2 days, 0:09:44, time: 83.26, loss: 0.00029862, loss_pre: 0.00000130, loss_aux0: 0.00016878, loss_aux1: 0.00009968, loss_aux2: 0.00002846, loss_aux3: 0.00000040
INFO logger.py(52): iter: 45000/150000, lr: 0.002007, eta: 2 days, 0:08:23, time: 83.52, loss: 0.00030348, loss_pre: 0.00000115, loss_aux0: 0.00017103, loss_aux1: 0.00010156, loss_aux2: 0.00002941, loss_aux3: 0.00000033
INFO logger.py(52): iter: 45050/150000, lr: 0.002006, eta: 2 days, 0:07:03, time: 83.47, loss: 0.00030164, loss_pre: 0.00000109, loss_aux0: 0.00017029, loss_aux1: 0.00010013, loss_aux2: 0.00002980, loss_aux3: 0.00000033
INFO logger.py(52): iter: 45100/150000, lr: 0.002005, eta: 2 days, 0:05:42, time: 83.24, loss: 0.00029908, loss_pre: 0.00000146, loss_aux0: 0.00016886, loss_aux1: 0.00009881, loss_aux2: 0.00002957, loss_aux3: 0.00000038
INFO logger.py(52): iter: 45150/150000, lr: 0.002004, eta: 2 days, 0:04:22, time: 83.40, loss: 0.00030068, loss_pre: 0.00000104, loss_aux0: 0.00017032, loss_aux1: 0.00010000, loss_aux2: 0.00002895, loss_aux3: 0.00000037
INFO logger.py(52): iter: 45200/150000, lr: 0.002004, eta: 2 days, 0:03:02, time: 83.51, loss: 0.00030403, loss_pre: 0.00000140, loss_aux0: 0.00017141, loss_aux1: 0.00010064, loss_aux2: 0.00003018, loss_aux3: 0.00000041
INFO logger.py(52): iter: 45250/150000, lr: 0.002003, eta: 2 days, 0:01:41, time: 83.37, loss: 0.00030358, loss_pre: 0.00000173, loss_aux0: 0.00017191, loss_aux1: 0.00010051, loss_aux2: 0.00002912, loss_aux3: 0.00000031
INFO logger.py(52): iter: 45300/150000, lr: 0.002002, eta: 2 days, 0:00:20, time: 83.33, loss: 0.00030002, loss_pre: 0.00000134, loss_aux0: 0.00017032, loss_aux1: 0.00009917, loss_aux2: 0.00002868, loss_aux3: 0.00000051
INFO logger.py(52): iter: 45350/150000, lr: 0.002001, eta: 1 day, 23:59:00, time: 83.58, loss: 0.00030427, loss_pre: 0.00000155, loss_aux0: 0.00017105, loss_aux1: 0.00010144, loss_aux2: 0.00002986, loss_aux3: 0.00000038
INFO logger.py(52): iter: 45400/150000, lr: 0.002000, eta: 1 day, 23:57:40, time: 83.52, loss: 0.00030417, loss_pre: 0.00000098, loss_aux0: 0.00017220, loss_aux1: 0.00010182, loss_aux2: 0.00002886, loss_aux3: 0.00000031
INFO logger.py(52): iter: 45450/150000, lr: 0.001999, eta: 1 day, 23:56:19, time: 83.42, loss: 0.00030224, loss_pre: 0.00000082, loss_aux0: 0.00017038, loss_aux1: 0.00010065, loss_aux2: 0.00003012, loss_aux3: 0.00000028
INFO logger.py(52): iter: 45500/150000, lr: 0.001998, eta: 1 day, 23:54:59, time: 83.37, loss: 0.00029685, loss_pre: 0.00000148, loss_aux0: 0.00016972, loss_aux1: 0.00009710, loss_aux2: 0.00002815, loss_aux3: 0.00000040
INFO logger.py(52): iter: 45550/150000, lr: 0.001997, eta: 1 day, 23:53:38, time: 83.12, loss: 0.00030367, loss_pre: 0.00000129, loss_aux0: 0.00017177, loss_aux1: 0.00009974, loss_aux2: 0.00003056, loss_aux3: 0.00000031
INFO logger.py(52): iter: 45600/150000, lr: 0.001997, eta: 1 day, 23:52:17, time: 83.37, loss: 0.00030221, loss_pre: 0.00000099, loss_aux0: 0.00017057, loss_aux1: 0.00010133, loss_aux2: 0.00002899, loss_aux3: 0.00000034
INFO logger.py(52): iter: 45650/150000, lr: 0.001996, eta: 1 day, 23:50:56, time: 83.30, loss: 0.00029999, loss_pre: 0.00000092, loss_aux0: 0.00016889, loss_aux1: 0.00010053, loss_aux2: 0.00002923, loss_aux3: 0.00000042
INFO logger.py(52): iter: 45700/150000, lr: 0.001995, eta: 1 day, 23:49:36, time: 83.51, loss: 0.00030480, loss_pre: 0.00000115, loss_aux0: 0.00017208, loss_aux1: 0.00010135, loss_aux2: 0.00002994, loss_aux3: 0.00000028
INFO logger.py(52): iter: 45750/150000, lr: 0.001994, eta: 1 day, 23:48:15, time: 83.43, loss: 0.00030246, loss_pre: 0.00000106, loss_aux0: 0.00017044, loss_aux1: 0.00010074, loss_aux2: 0.00002984, loss_aux3: 0.00000038
INFO logger.py(52): iter: 45800/150000, lr: 0.001993, eta: 1 day, 23:46:55, time: 83.33, loss: 0.00029826, loss_pre: 0.00000134, loss_aux0: 0.00016861, loss_aux1: 0.00009860, loss_aux2: 0.00002924, loss_aux3: 0.00000046
INFO logger.py(52): iter: 45850/150000, lr: 0.001992, eta: 1 day, 23:45:34, time: 83.61, loss: 0.00030538, loss_pre: 0.00000106, loss_aux0: 0.00017260, loss_aux1: 0.00010121, loss_aux2: 0.00002993, loss_aux3: 0.00000058
INFO logger.py(52): iter: 45900/150000, lr: 0.001991, eta: 1 day, 23:44:14, time: 83.33, loss: 0.00030398, loss_pre: 0.00000128, loss_aux0: 0.00017081, loss_aux1: 0.00010138, loss_aux2: 0.00002960, loss_aux3: 0.00000090
INFO logger.py(52): iter: 45950/150000, lr: 0.001991, eta: 1 day, 23:42:53, time: 83.56, loss: 0.00030061, loss_pre: 0.00000096, loss_aux0: 0.00017041, loss_aux1: 0.00009921, loss_aux2: 0.00002962, loss_aux3: 0.00000039
INFO logger.py(52): iter: 46000/150000, lr: 0.001990, eta: 1 day, 23:41:33, time: 83.46, loss: 0.00030364, loss_pre: 0.00000295, loss_aux0: 0.00016952, loss_aux1: 0.00010075, loss_aux2: 0.00002997, loss_aux3: 0.00000045
INFO logger.py(52): iter: 46050/150000, lr: 0.001989, eta: 1 day, 23:40:13, time: 83.58, loss: 0.00029999, loss_pre: 0.00000107, loss_aux0: 0.00016928, loss_aux1: 0.00009935, loss_aux2: 0.00002984, loss_aux3: 0.00000045
INFO logger.py(52): iter: 46100/150000, lr: 0.001988, eta: 1 day, 23:38:52, time: 83.45, loss: 0.00030163, loss_pre: 0.00000218, loss_aux0: 0.00016981, loss_aux1: 0.00010011, loss_aux2: 0.00002921, loss_aux3: 0.00000032
INFO logger.py(52): iter: 46150/150000, lr: 0.001987, eta: 1 day, 23:37:32, time: 83.53, loss: 0.00030411, loss_pre: 0.00000121, loss_aux0: 0.00017053, loss_aux1: 0.00010123, loss_aux2: 0.00003006, loss_aux3: 0.00000108
INFO logger.py(52): iter: 46200/150000, lr: 0.001986, eta: 1 day, 23:36:12, time: 83.69, loss: 0.00030011, loss_pre: 0.00000098, loss_aux0: 0.00016882, loss_aux1: 0.00009999, loss_aux2: 0.00002978, loss_aux3: 0.00000054
INFO logger.py(52): iter: 46250/150000, lr: 0.001985, eta: 1 day, 23:34:51, time: 83.24, loss: 0.00030160, loss_pre: 0.00000098, loss_aux0: 0.00016927, loss_aux1: 0.00010116, loss_aux2: 0.00002970, loss_aux3: 0.00000049
INFO logger.py(52): iter: 46300/150000, lr: 0.001985, eta: 1 day, 23:33:31, time: 83.50, loss: 0.00030482, loss_pre: 0.00000143, loss_aux0: 0.00017021, loss_aux1: 0.00010200, loss_aux2: 0.00003083, loss_aux3: 0.00000035
INFO logger.py(52): iter: 46350/150000, lr: 0.001984, eta: 1 day, 23:32:10, time: 83.52, loss: 0.00030408, loss_pre: 0.00000111, loss_aux0: 0.00017039, loss_aux1: 0.00010202, loss_aux2: 0.00003007, loss_aux3: 0.00000049
INFO logger.py(52): iter: 46400/150000, lr: 0.001983, eta: 1 day, 23:30:50, time: 83.56, loss: 0.00030009, loss_pre: 0.00000137, loss_aux0: 0.00016943, loss_aux1: 0.00009996, loss_aux2: 0.00002900, loss_aux3: 0.00000033
INFO logger.py(52): iter: 46450/150000, lr: 0.001982, eta: 1 day, 23:29:29, time: 83.39, loss: 0.00030261, loss_pre: 0.00000107, loss_aux0: 0.00016924, loss_aux1: 0.00010090, loss_aux2: 0.00003112, loss_aux3: 0.00000028
INFO logger.py(52): iter: 46500/150000, lr: 0.001981, eta: 1 day, 23:28:08, time: 83.36, loss: 0.00030534, loss_pre: 0.00000107, loss_aux0: 0.00016925, loss_aux1: 0.00010376, loss_aux2: 0.00003080, loss_aux3: 0.00000046
INFO logger.py(52): iter: 46550/150000, lr: 0.001980, eta: 1 day, 23:26:48, time: 83.38, loss: 0.00029968, loss_pre: 0.00000107, loss_aux0: 0.00016872, loss_aux1: 0.00009909, loss_aux2: 0.00003043, loss_aux3: 0.00000036
INFO logger.py(52): iter: 46600/150000, lr: 0.001979, eta: 1 day, 23:25:27, time: 83.61, loss: 0.00030338, loss_pre: 0.00000372, loss_aux0: 0.00016950, loss_aux1: 0.00010051, loss_aux2: 0.00002930, loss_aux3: 0.00000035
**

**Comment by zjwzcx** [comment]
I am very confused. Am I testing the convergence of the model the wrong way? Or is my number of iterations not overfitting the model? @CoinCheung 
I'm very anxious to complete this task, if you are interested, I can give you access to my private repository. Thank you!

**Comment by CoinCheung** [comment]
Hi,

It may relate to your training configuration. From the first look at the description, my intuition is overfitting. Why do you use training set with only 1 image? Did you test your model on the identical image?

**Comment by zjwzcx** [comment]
> Hi,
> 
> It may relate to your training configuration. From the first look at the description, my intuition is overfitting. Why do you use training set with only 1 image? Did you test your model on the identical image?

 _**Why do you use training set with only 1 image?**_ 
My dataset is a little dirty (inaccurate labels, ...), so I want to make sure BiSeNet can learn the feature from this kind of dataset. If the model can be trained long enough on 1 image and visualization performs well, I‘ll think the model can work on my dataset.

 _** Did you test your model on the identical image**_ 
Yes! I trained,tested and visualized the model on the identical image.

**Comment by CoinCheung** [comment]
I doubt whether training on one image would work at all, though I do not have such experience. Maybe you can reduce the training iterations. For example, you can train for only 100 iterations, and see the result.

**Comment by zjwzcx** [comment]
> I doubt whether training on one image would work at all, though I do not have such experience. Maybe you can reduce the training iterations. For example, you can train for only 100 iterations, and see the result.

Thanks for your suggestionm I'll try it soon. BTW, How many iterations do we typically choose for BiSeNet on CityScapes dataset (5k training samples)? I noticed that the maximum number of iterations in the config file is 150k, which confused me.

**Comment by zjwzcx** [comment]
I did an experiment: I trained for 20 iterations on the full dataset and found that the results based on the 5th~7th iteration were ideal.
My previous thought was really, really stupid.  :(

Looking forward to your answer to my last question (How many iterations do we typically choose for BiSeNet on CityScapes dataset). I'll close this issue right away.

**Comment by CoinCheung** [comment]
Hi,

Good to know you have made it work.

Cityscapes dataset has only  2975 training images of 19 categories for semantic segmentation. The images are as large as 1024x2048. In general, we do not directly use the whole large images for training. Instead, we usually crop patches from the original images as input (for example, a cropsize of 768x768). 

As for training iterations, it depends on the specific models and regularizations. Models with big volume are good at fitting the dataset, thus we can use fewer training iterations (for example, 40k-80k). For the light-weighted models, we always use more training iterations(150k for bisenet) because they are likely to under-fit on the dataset. Additionally, if you add strong regularizations(such as augmentations) to the training samples, you can prolong the training iterations properly. A commonly accepted method to determine these training parameters is using "grid search".

Hope this is helpful :)

**Comment by zjwzcx** [comment]
> 

That's very helpful for me! Thanks for your detailed explanation!
